libraries/Microsoft.Bot.Builder.Azure/CosmosDbStorage.cs

```diff
@@ -0,0 +1,236 @@
+                    if (item is IStoreItem storeItem)
+                    {
+                        storeItem.eTag = doc.ETag;
+                    }
+
+                    // doc.Id cannot be used since it is escaped, read it from RealId property instead
+                    storeItems[doc.ReadlId] = item;
+                }
+            }
+
+            return storeItems;
+        }
+
+        /// <summary>
+        /// Persist the specified StoreItems into the CosmosDB Collection.
+        /// </summary>
+        /// <param name="changes">Items to persist.</param>
+        /// <returns></returns>
+        public async Task Write(StoreItems changes)
+        {
+            if (changes == null)
+            {
+                throw new ArgumentNullException(nameof(changes), "Please provide a StoreItems with changes to persist.");
+            }
+
+            var collectionLink = _collectionLink.Value;
+            foreach (var change in changes)
+            {
+                var json = JObject.FromObject(change.Value, _jsonSerializer);
+                json.Remove("eTag");
+                var documentChange = new DocumentStoreItem
+                {
+                    Id = SanitizeKey(change.Key),
+                    ReadlId = change.Key,
+                    Document = json
+                };
+
+                string eTag = (change.Value as IStoreItem)?.eTag;
+                if (eTag == null || eTag == "*")
+                {
+                    // if new item or * then insert or replace unconditionaly
+                    await _client.UpsertDocumentAsync(collectionLink, documentChange, disableAutomaticIdGeneration: true).ConfigureAwait(false);
+                }
+                else if (eTag.Length > 0)
+                {
+                    // if we have an etag, do opt. concurrency replace
+                    var uri = UriFactory.CreateDocumentUri(_databaseId, _collectionId, documentChange.Id);
+                    var ac = new AccessCondition { Condition = eTag, Type = AccessConditionType.IfMatch };
+                    await _client.ReplaceDocumentAsync(uri, documentChange, new RequestOptions { AccessCondition = ac }).ConfigureAwait(false);
+                }
+                else
+                {
+                    throw new Exception("etag empty");
+                }
+            }
+        }
+
+        private string EnsureCollectionExist()
+        {
+            _client.CreateDatabaseIfNotExistsAsync(new Database { Id = _databaseId })
+                .Wait();
+
+            var task = _client.CreateDocumentCollectionIfNotExistsAsync(UriFactory.CreateDatabaseUri(_databaseId), new DocumentCollection { Id = _collectionId });
+            task.Wait();
+
+            return task.Result.Resource.SelfLink;
+        }
+
+        private static Lazy<Dictionary<char, string>> _badChars = new Lazy<Dictionary<char, string>>(() =>
+        {
+            char[] badChars = new char[] { '\\', '?', '/', '#', ' ' };
```

Comment:
Where does this list come from? Could you add a link, and/or describe? 

For example, is this list from CosmosDB? or from JSON? or from HTML? or somewhere else?